### ISSUE TITLE
Wistar roller blinds - not fully functional

### ISSUE THREAD

**Issue by mihsu81** [make-all/tuya-local#320]
Hi @make-all,

I've tried using today the Wistar roller blinds profile but it's not fully functional. I've encountered 3 apparent issues:

Issue 1. The Reversed switch (dpID 103) being enabled actually means normal operation, not reversed operation. I couldn't find an option to revert it in the device profile.
![image](https://user-images.githubusercontent.com/40715497/209403585-7d65eff1-2c7f-4845-9ed7-a9596483c464.png)

Issue 2. Only one command is available: Close or Open, depending on the status of the Reversed switch.
![image](https://user-images.githubusercontent.com/40715497/209403060-fe5331ac-fab8-4b3e-af1e-9845c9624345.png)

Issue 3. The current open/close position of the motor is not reflected in HA. Unless I set a position, it never changes.
![image](https://user-images.githubusercontent.com/40715497/209402990-68045e42-78ff-4592-9406-6e19cccf8b3d.png)

P.S. The response of the motor is so much better now than it was months ago. Congratulations. 👍🏼

**Comment by make-all** [comment]
I think the problem of open being disabled when the blind is closed may also be linked to the reversal of the reverse setting.

Position can only report what the device is reporting.  If the device is only using the position as input, then it will not change unless explicitly set.

**Comment by make-all** [comment]
I just made a change that flips the meaning of reversed c3d78509bc3c.  After this, it will be clearer what problems are separate issues that still remain.

**Comment by mihsu81** [comment]
Thanks @make-all. https://github.com/make-all/tuya-local/commit/c3d78509bc3cf77099ea521a25877d0ea75ee8b9 fixed the reverse switch.
But issues 2 and 3 remain.

**Comment by mihsu81** [comment]
Hi @make-all,
Here are de device diagnostics.
[tuya_local-bf945b4460e06d653429729f892ca577-Living Room Shutter-ce09bf37ebf974b74724daba715db421.zip](https://github.com/make-all/tuya-local/files/10303948/tuya_local-bf945b4460e06d653429729f892ca577-Living.Room.Shutter-ce09bf37ebf974b74724daba715db421.zip)

**Comment by make-all** [comment]
One potential issue is that I named an attribute "state", which might interfere with Home Assistant's expectation of what the main state for a cover device is.  I don't really know what the extent of the effect of this would be, but everything else looks normal.

Another possibility - have you checked that the position sensor is working? (eg via Tuya cloud).  The current position is being reported as 0 (fully closed), which if it is not correct will prevent the close function from working, since Home Assistant seems to disable the open/close controls when it is at the respective end..

**Comment by mihsu81** [comment]
Yes, The position is reported correctly in the Tuya app, but also in the localtuya integration which I'm currently using for the shutters.

**Comment by make-all** [comment]
Position in the diagnostics is 0 (fully open).  This is what is being read from the device.

Can you take the diagnostics when it is stopped somewhere in the middle, as per the screenshot for Issue 2.  Also if you can send diagnostics when reproducing Issue 3.

**Comment by make-all** [comment]
I am not sure that Issue 3 is a bug, it may be just the way that Home Assistant works.  Is current position separately listed under Attributes?

**Comment by mihsu81** [comment]
> Position in the diagnostics is 0 (fully open). This is what is being read from the device.

Position 0 is fully closed, not fully open. DPID 112 should reflect the current position but as far as I can see it remains 0.
DPID 102, which is "set position" does change even when only using the arrow to move the motor.
[tuya_local-bf945b4460e06d653429729f892ca577-Living Room Shutter-ce09bf37ebf974b74724daba715db421.json (1).txt](https://github.com/make-all/tuya-local/files/10325861/tuya_local-bf945b4460e06d653429729f892ca577-Living.Room.Shutter-ce09bf37ebf974b74724daba715db421.json.1.txt)

**Comment by make-all** [comment]
It seems the position sensor in this device is not functioning, and that is why you are not seeing the position reflected.  It seemed to be working in #228, so is there a possibility that it has broken since then?

**Comment by make-all** [comment]
Please try removing the device, and re-adding it with the new profile I just added (wistar_roller_blind_nopos.yaml)

This ignores the current_position dp, and just uses position.

**Comment by mihsu81** [comment]
The issue seems to be caused by how often Tuya Local checks the value of dpID 112.
Tuya Local shows dpID 112 as 100 (fuly open) instead of 79, but localTuya and iot.tuya.com read it properly as 79.
Tuya Local showed dpID 112 as 79 only after reloading the integration. I've waited about 5 minutes before reloading the integration and the value hadn't been updated.
![image](https://user-images.githubusercontent.com/40715497/210137313-ad788b6d-22a9-4883-80dc-096dbd3f8776.png)
![image](https://user-images.githubusercontent.com/40715497/210137321-62fbdf38-febb-4e03-86d0-9713f7946b51.png)
![image](https://user-images.githubusercontent.com/40715497/210137336-8ca84219-e669-45e4-9bc6-55f76f9c1685.png)
![image](https://user-images.githubusercontent.com/40715497/210137459-b9e206d3-4dd5-4649-979e-39c8d97fc601.png)

**Comment by make-all** [comment]
Tuya-local doesn't check individual dps separately. It gets full updates from the device on each poll.

**Comment by mihsu81** [comment]
I don't know if localtuya or iot.tuya check the dpIDs individually, I was just mentioning my findings. No disrespect whatsoever to your work.
I forgot to mention in my previous post that I had tried the profile wistar_roller_blind_nopos but it still wasn't displaying the corect position.
However I've upgraded now to v2023.1.0 and tried again with the profile wistar_roller_blind_nopos and the position of the motor gets updated properly. But it does take a long time between updates ~ 15 seconds. Is that configurable?
Thanks a lot for your work.

**Comment by make-all** [comment]
The update interval seems to depend on the platform, I'm not sure where in Home Assistant this is determined and whether it is configurable at all.  I am however working on moving to persistent connections instead of polling, which is how localtuya manages to reflect state changes so quickly.  But there is a big learning curve here with how Python and Home Assistant manage threads and asynchronous I/O.  Localtuya forked their own version of the pytuya library after the upstream developer abandoned it, and have converted their version to use asyncio, but tinytuya, which is the maintained fork I chose to use still does not support asyncio, so it isn't so easy to fit into Home Assistant's async architecture.

**Comment by mihsu81** [comment]
Thanks a lot for the explanation. It's all much clearer now. 👍

**Comment by mihsu81** [comment]
@make-all I've started using the WiFi motors with version 2023.1.2 and it's working really well with the persistent connections you added in 2023.1.1.
Congratulations on the new implementation. 👍
